### PR TITLE
Add allowed-plugin param to enable plugins allow-listing

### PR DESCRIPTION
**Pull request by jakubm-canva**
Protection from malicious plugins - An allowed-plugins command line param accepting an allow-list of regular expression strings. e.g.: ^github.com/buildkite/.*$. If the param is used and a job arrives with BUILDKITE_PLUGINS containing a plugin that doesn't match an entry in the list then the job is rejected.

Work is a part of the https://github.com/buildkite/agent/issues/2368

### CODE REVIEW

**Comment by jakubm-canva** [comment]
Ready to review @david-poirier & @CheeseStick

File: agent/run_job.go [hunk 1]
```diff
@@ -116,17 +118,32 @@ func (r *JobRunner) Run(ctx context.Context) error {
...
 		return nil
 	}
+	// Unmarshal plugins and validate the plugins if the list of allowed plugins is set.
+	// Erros in json.Unmarshal will by-pass the plugin verification.
+	var ps pipeline.Plugins
+	if err := json.Unmarshal([]byte(job.Env["BUILDKITE_PLUGINS"]), &ps); len(job.Env["BUILDKITE_PLUGINS"]) > 0 && err == nil {
+		for _, plugin := range ps {
+			err = validateJobValue(r.conf.AgentConfiguration.AllowedPlugins, plugin.Source)
+			if err != nil {
+				errStr := fmt.Sprintf("Plugin %s not allowed (BUILDKITE_PLUGINS) - %s", job.Env["BUILDKITE_PLUGINS"], err)
```
[the remark this answers is not in the document]
Comment (jakubm-canva) [reply agent/run_job.go:138]: Oh yea, you right. Thanks

File: agent/run_job.go [hunk 2]
```diff
@@ -116,17 +118,32 @@ func (r *JobRunner) Run(ctx context.Context) error {
...
-		r.logStreamer.Process([]byte(fmt.Sprintf("%s", err)))
-		r.logger.Error("%s", err)
+		errStr := fmt.Sprintf("Repo %s not allowed (BUILDKITE_REPO) - %s", job.Env["BUILDKITE_REPO"], err)
+		r.logStreamer.Process([]byte(errStr))
+		r.logger.Error("%s", errStr)
 		exit.Status = -1
 		exit.SignalReason = SignalReasonAgentRefused
 		return nil
 	}
+	// Unmarshal plugins and validate the plugins if the list of allowed plugins is set.
```
[the remark this answers is not in the document]
Comment (jakubm-canva) [reply agent/run_job.go:131]: Done.

**Comment by jakubm-canva** [comment]
@triarius - Hey. Would it be possible to take a look on this change. Thank you, Jakub

File: agent/job_runner.go [hunk 3]
```diff
@@ -570,17 +570,19 @@ func (w LogWriter) Write(bytes []byte) (int, error) {
...
 
-	for _, allowedRepo := range allowedRepos {
-		if match, _ := regexp.MatchString(allowedRepo, pipelineRepo); match {
+	for _, at := range allowedPatterns {
+		if match, _ := regexp.MatchString(at, jobValue); match {
 			return nil
 		}
 	}
-	return fmt.Errorf("repository %s isn't allowed", pipelineRepo)
+	return fmt.Errorf("%s has no match in %s", jobValue, allowedPatterns)
```
Comment (moskyb) [inline agent/job_runner.go:585]: i wonder if it'd be nice to say where the list of patterns came from - ie have the error message be something like:
```
github.com/crimes/evil-buildkite-plugin has no match in allowed plugins: [github.com/buildkite/.*]
```

**Review by moskyb — CHANGES_REQUESTED** [review]
looking good! there's a couple of small changes i'd like to see, but other than that this is looking great 🎉

File: agent/run_job.go [hunk 4]
```diff
@@ -116,17 +118,34 @@ func (r *JobRunner) Run(ctx context.Context) error {
...
+					errStr := fmt.Sprintf("Plugin %s not allowed (BUILDKITE_PLUGINS) - %s", plugin.Source, err)
+					r.logStreamer.Process([]byte(errStr))
+					r.logger.Error("%s", errStr)
+					exit.Status = -1
+					exit.SignalReason = SignalReasonAgentRefused
+					return nil
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jakubm-canva) [reply agent/run_job.go]: Done. 
Example error:
`Plugin ssh://git@github.com/arromer/github-fetch-buildkite-plugin.git has no match in allowed patterns [^you-shall-not-pass.*]`

**Comment by jakubm-canva** [comment]
@moskyb PTAL

File: clicommand/agent_start.go [hunk 5]
```diff
@@ -851,6 +858,7 @@ var AgentStartCommand = cli.Command{
 			PluginValidation:                !cfg.NoPluginValidation,
 			LocalHooksEnabled:               !cfg.NoLocalHooks,
 			AllowedRepositories:             cfg.AllowedRepositories,
+			AllowedPlugins:                  cfg.AllowedPlugins,
```
Comment (moskyb) [inline clicommand/agent_start.go:861]: like in the other PR, we should compile these regexes in this file, and then pass them to the `JobRunner` as a `[]*regexp.Regexp`. probably a good idea to backport this behaviour to `allowed-repos` as well

**Review by moskyb — CHANGES_REQUESTED** [review]
one relatively small thing, but other than that LGTM!

File: clicommand/agent_start.go (hunk 5 above)
Comment (jakubm-canva) [reply clicommand/agent_start.go:861]: Like in the other PR. What other PR?

**Pull request closed without merging**